File: web/src/app/shared/models/reservation.ts
import { Livre } from "./livre";
import { Member } from "./member";

export class Reservation {
  private _idReservation?: number;
  private _dateReservation: Date;
  private _memberId?: number;
  private _livreId?: number;
  member?: Member;  // Add this line
  livre?: Livre;
  deletedAt?:Date;

  constructor(
      idReservation?: number,
      memberId?: number,
      livreId?: number,
      dateReservation: Date = new Date() // Default to current date if not provided
  ) {
      this._idReservation = idReservation;
      this._memberId = memberId;
      this._livreId = livreId;
      this._dateReservation = dateReservation;
  }

  // Getters and Setters
  get idReservation(): number | undefined {
      return this._idReservation;
  }

  set idReservation(idReservation: number) {
      this._idReservation = idReservation;
  }

  get dateReservation(): Date {
      return this._dateReservation;
  }

  set dateReservation(dateReservation: Date) {
      this._dateReservation = dateReservation;
  }

  get memberId(): number | undefined {
      return this._memberId;
  }

  set memberId(memberId: number | undefined) {
      this._memberId = memberId;
  }

  get livreId(): number | undefined {
      return this._livreId;
  }

  set livreId(livreId: number | undefined) {
      this._livreId = livreId;
  }
}

File: back/src/main/java/org/bibliotheque/bibliotheque/modele/entity/Reservation.java
package org.bibliotheque.bibliotheque.modele.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.Date;

@AllArgsConstructor
@NoArgsConstructor
@Data
@SuperBuilder
@Entity
@Table
public class Reservation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_reservation")
    private Integer idReservation;

    @NotNull
    @NotBlank(message = "La date de reservation ne doit pas être vide")
    private Date dateReservation;

    @CreationTimestamp
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    // Foreign keys
    @Column(name = "member_id")
    private Integer memberId;

    @Column(name = "livre_id")
    private Integer livreId;

    // Relationships
    @JsonBackReference("reservation_member")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", insertable = false, updatable = false)
    private Member member;

    @JsonBackReference("reservation_livre")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "livre_id", insertable = false, updatable = false)
    private Livre livre;

    @Override
    public String toString() {
        return "Reservation{" +
                "idReservation=" + idReservation +
                ", dateReservation=" + dateReservation +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                ", deletedAt=" + deletedAt +
                ", memberId=" + memberId +
                ", livreId=" + livreId +
                '}';
    }
}
